### ISSUE TITLE
defaultValue: Any, allowNull: false should not throw not-null violation errors.

### ISSUE THREAD

**Issue by mirek** [sequelize/sequelize#4244]
Current behaviour for columns defined with:
```
{ defaultValue: Any, allowNull: false }
```
- defines sql column with default value and not-null constraint (good)
- allows for creating objects where field has been omitted in `Model.create({ ... })` (good)
- throws non-null violation if field is set to `undefined` in `Model.create({ ... })` (wrong)
- throws non-null violation if field is set to `null` in `Model.create({ ... })` (wrong)

Current workaround is to define special setters for each column defined with default value and not-null constraint:

```
setterMethods: {
  'myfield': function (value) {
    this.setDataValue('myfield', value != null ? value : myfield_default);
  },
}
```

Is this behaviour intentional and this kind of workaround recommended solution to this problem?

**Comment by janmeier** [comment]
I believe this is actually intential, even though I can see why its annoying in this case.

Adding the default value when the column is null would _not_ be a good idea in my mind - it would be too magical to replace what the user passed in with the default value. When you try to insert null, you should get an error that that's not allowed :)

As for undefined, there might be an argument for replacing that, but then again the distinct between null and undefined is so small that I think it would be confusing to replace one but not the other.

So in conclusion I see why this might be annoying, but I think what you're doing her eis the best solution

**Comment by mirek** [comment]
Ok, fair enough, just for reference this is probably the best way to handle it (via `DEFAULT` literal):

```
/* eslint-env node */

import Sequelize from 'sequelize';
import _ from 'lodash';

/**
 * Setter method transforms null and undefined values into DEFAULT literal.
 * @param {any} value
 */
function autoDefault(value) {
  if (value != null) {
    this.setDataValue('value', value);
  } else {
    this.setDataValue('value', Sequelize.literal('DEFAULT'))
  }
}

async function main() {
  const sequelize = new Sequelize('auto-default-test', null, null, { dialect: 'postgres', logging: null });

  const Foo = sequelize.define('Foo', {
    'value': { type: Sequelize.TEXT, defaultValue: 'Hello world!', allowNull: false },
  }, {
    tableName: 'foos',
    setterMethods: {
      value: autoDefault // Set "value" attribute to default value for null/undefined instead of throwing non-null violation.
    }
  });

  async function dump() {
    console.log(await sequelize.query('SELECT "id", "value" FROM "foos"', { type: Sequelize.QueryTypes.SELECT }));
  }

  await sequelize.sync();

  await Foo.destroy({ where: true });

  // Will use default for all 3 cases:
  await Foo.create();
  await Foo.create({ value: undefined });
  await Foo.create({ value: null });

  await Foo.create({ value: 'foo' });

  await dump();
  // [ { id: 25, value: 'Hello world!' },
  //   { id: 26, value: 'Hello world!' },
  //   { id: 27, value: 'Hello world!' },
  //   { id: 28, value: 'foo' } ]

  let foo = await Foo.findOne({ where: { value: 'foo' } })
  await foo.updateAttributes({ value: null }); // Will use default as well.

  await dump();
  // [ { id: 25, value: 'Hello world!' },
  //   { id: 26, value: 'Hello world!' },
  //   { id: 27, value: 'Hello world!' },
  //   { id: 28, value: 'Hello world!' } ]

}

main().catch(err => console.error(err.stack));
```
